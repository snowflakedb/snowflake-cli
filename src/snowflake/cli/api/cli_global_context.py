# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field, replace
from functools import wraps
from pathlib import Path
from typing import TYPE_CHECKING, Iterator, Optional

from snowflake.cli.api.connections import ConnectionContext, OpenConnectionCache
from snowflake.cli.api.exceptions import MissingConfiguration
from snowflake.cli.api.metrics import CLIMetrics
from snowflake.cli.api.output.formats import OutputFormat
from snowflake.cli.api.rendering.jinja import CONTEXT_KEY
from snowflake.connector import SnowflakeConnection

if TYPE_CHECKING:
    from snowflake.cli.api.project.definition_manager import DefinitionManager
    from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition

_CONNECTION_CACHE = OpenConnectionCache()


@dataclass
class _CliGlobalContextManager:
    connection_context: ConnectionContext = field(default_factory=ConnectionContext)
    connection_cache: OpenConnectionCache = (
        _CONNECTION_CACHE  # by default, use global cache
    )

    output_format: OutputFormat = OutputFormat.TABLE
    silent: bool = False
    verbose: bool = False
    experimental: bool = False
    enable_tracebacks: bool = True

    metrics: CLIMetrics = field(default_factory=CLIMetrics)

    project_path_arg: str | None = None
    project_is_optional: bool = True
    project_env_overrides_args: dict[str, str] = field(default_factory=dict)

    # FIXME: this property only exists to help implement
    # nativeapp_definition_v2_to_v1 and single_app_and_package.
    # Consider changing the way this calculation is provided to commands
    # in order to remove this logic (then make project_definition a non-cloned @property)
    override_project_definition: ProjectDefinition | None = None

    _definition_manager: DefinitionManager | None = None

    # which properties invalidate our current DefinitionManager?
    DEFINITION_MANAGER_DEPENDENCIES = [
        "project_path_arg",
        "project_is_optional",
        "project_env_overrides_args",
    ]

    def reset(self):
        self.__init__()

    def clone(self) -> _CliGlobalContextManager:
        return replace(
            self,
            connection_context=self.connection_context.clone(),
            project_env_overrides_args=self.project_env_overrides_args.copy(),
            metrics=self.metrics.clone(),
        )

    def __setattr__(self, prop, val):
        if prop in self.DEFINITION_MANAGER_DEPENDENCIES:
            self._clear_definition_manager()

        super().__setattr__(prop, val)

    @property
    def project_definition(self) -> ProjectDefinition | None:
        if self.override_project_definition:
            return self.override_project_definition

        return self._definition_manager_or_raise().project_definition

    @property
    def project_root(self) -> Path:
        return Path(self._definition_manager_or_raise().project_root)

    @property
    def template_context(self) -> dict:
        return self._definition_manager_or_raise().template_context

    @property
    def connection(self) -> SnowflakeConnection:
        """
        Returns a connection for our configured context from the configured cache.
        By default, this is the global _CONNECTION_CACHE. If a matching connection
        does not already exist, creates a new connection and caches it.
        """
        self.connection_context.validate_and_complete()
        return self.connection_cache[self.connection_context]

    def _definition_manager_or_raise(self) -> DefinitionManager:
        """
        (Re-)parses project definition based on project args (project_path_arg and
        project_env_overrides_args). If we cannot provide a project definition
        (i.e. no snowflake.yml) and require one, raises MissingConfiguration.
        """
        from snowflake.cli.api.project.definition_manager import DefinitionManager

        # don't need to re-parse definition if we already have one
        if not self._definition_manager:
            dm = DefinitionManager(
                self.project_path_arg,
                {CONTEXT_KEY: {"env": self.project_env_overrides_args}},
            )
            if not dm.has_definition_file and not self.project_is_optional:
                raise MissingConfiguration(
                    "Cannot find project definition (snowflake.yml). Please provide a path to the project or run this command in a valid project directory."
                )
            self._definition_manager = dm

        return self._definition_manager

    def _clear_definition_manager(self):
        """
        Force re-calculation of definition_manager and its dependent attributes
        (template_context, project_definition, and project_root).
        """
        self._definition_manager = None


class _CliGlobalContextAccess:
    def __init__(self, manager: _CliGlobalContextManager):
        self._manager = manager

    @property
    def connection(self) -> SnowflakeConnection:
        return self._manager.connection

    @property
    def connection_context(self) -> ConnectionContext:
        return self._manager.connection_context

    @property
    def enable_tracebacks(self) -> bool:
        return self._manager.enable_tracebacks

    @property
    def metrics(self) -> CLIMetrics:
        return self._manager.metrics

    @property
    def output_format(self) -> OutputFormat:
        return self._manager.output_format

    @property
    def verbose(self) -> bool:
        return self._manager.verbose

    @property
    def experimental(self) -> bool:
        return self._manager.experimental

    @property
    def project_definition(self) -> ProjectDefinition | None:
        return self._manager.project_definition

    @property
    def project_root(self) -> Path | None:
        return self._manager.project_root

    @property
    def template_context(self) -> dict:
        return self._manager.template_context

    @property
    def silent(self) -> bool:
        if self._should_force_mute_intermediate_output:
            return True
        return self._manager.silent

    @property
    def _should_force_mute_intermediate_output(self) -> bool:
        """Computes whether cli_console output should be muted."""
        return self._manager.output_format == OutputFormat.JSON

    @property
    def snow_api_root(
        self,
    ) -> Optional[
        object
    ]:  # Should be Optional[Root], but we need local import for performance reasons
        from snowflake.core import Root

        if self.connection:
            return Root(self.connection)
        else:
            return None


_CLI_CONTEXT_MANAGER: ContextVar[_CliGlobalContextManager | None] = ContextVar(
    "cli_context", default=None
)


def get_cli_context_manager() -> _CliGlobalContextManager:
    mgr = _CLI_CONTEXT_MANAGER.get()
    if not mgr:
        mgr = _CliGlobalContextManager()
        _CLI_CONTEXT_MANAGER.set(mgr)
    return mgr


def get_cli_context() -> _CliGlobalContextAccess:
    return _CliGlobalContextAccess(get_cli_context_manager())


def span(span_name: str):
    """
    Decorator to start a command metrics span that encompasses a whole function

    Must be used instead of directly calling @get_cli_context().metrics.span(span_name)
    as a decorator to ensure that the cli context is grabbed at run time instead of at
    module load time, which would not reflect forking
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with get_cli_context().metrics.span(span_name):
                return func(*args, **kwargs)

        return wrapper

    return decorator


@contextmanager
def fork_cli_context(
    connection_overrides: dict | None = None,
    project_env_overrides: dict[str, str] | None = None,
    project_is_optional: bool | None = None,
    project_path: str | None = None,
) -> Iterator[_CliGlobalContextAccess]:
    """
    Forks the global CLI context, making changes that are only visible
    (e.g. via get_cli_context()) while inside this context manager.

    Please note that environment variable changes are only visible through
    the project definition; os.getenv / os.environ / get_env_value are not
    affected by these new values.
    """
    old_manager = get_cli_context_manager()
    new_manager = old_manager.clone()
    token = _CLI_CONTEXT_MANAGER.set(new_manager)

    if connection_overrides:
        new_manager.connection_context.update(**connection_overrides)

    if project_env_overrides:
        new_manager.project_env_overrides_args.update(project_env_overrides)

    if project_is_optional is not None:
        new_manager.project_is_optional = project_is_optional

    if project_path:
        new_manager.project_path_arg = project_path

    yield _CliGlobalContextAccess(new_manager)
    _CLI_CONTEXT_MANAGER.reset(token)
