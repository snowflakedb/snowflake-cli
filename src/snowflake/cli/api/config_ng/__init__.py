# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Enhanced Configuration System - Next Generation (NG)

This package implements a simple, extensible configuration system with:
- Two-phase resolution: file sources use connection-level replacement,
  overlay sources (env/CLI) use field-level merging
- List-order precedence (explicit ordering in source list)
- Migration support (SnowCLI and SnowSQL compatibility)
- Complete resolution history tracking
- Read-only, immutable configuration sources
"""

from snowflake.cli.api.config_ng.core import (
    ConfigValue,
    ResolutionEntry,
    ResolutionHistory,
    SourceType,
    ValueSource,
)
from snowflake.cli.api.config_ng.presentation import ResolutionPresenter
from snowflake.cli.api.config_ng.resolution_logger import (
    check_value_source,
    explain_configuration,
    export_resolution_history,
    format_summary_for_display,
    get_resolution_summary,
    get_resolver,
    is_resolution_logging_available,
    show_all_resolution_chains,
    show_resolution_chain,
)
from snowflake.cli.api.config_ng.resolver import (
    ConfigurationResolver,
    ResolutionHistoryTracker,
)
from snowflake.cli.api.config_ng.sources import (
    CliConfigFile,
    CliEnvironment,
    CliParameters,
    ConnectionsConfigFile,
    ConnectionSpecificEnvironment,
    SnowSQLConfigFile,
    SnowSQLEnvironment,
    SnowSQLSection,
    get_merged_variables,
)

__all__ = [
    "check_value_source",
    "CliConfigFile",
    "CliEnvironment",
    "CliParameters",
    "ConfigurationResolver",
    "ConfigValue",
    "ConnectionsConfigFile",
    "ConnectionSpecificEnvironment",
    "explain_configuration",
    "export_resolution_history",
    "format_summary_for_display",
    "get_merged_variables",
    "get_resolution_summary",
    "get_resolver",
    "is_resolution_logging_available",
    "ResolutionEntry",
    "ResolutionHistory",
    "ResolutionHistoryTracker",
    "ResolutionPresenter",
    "show_all_resolution_chains",
    "show_resolution_chain",
    "SnowSQLConfigFile",
    "SnowSQLEnvironment",
    "SnowSQLSection",
    "SourceType",
    "ValueSource",
]
