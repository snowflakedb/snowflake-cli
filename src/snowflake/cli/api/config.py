# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
import os
import warnings
from contextlib import contextmanager
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

import tomlkit
from click import ClickException
from snowflake.cli.api.exceptions import (
    ConfigFileTooWidePermissionsError,
    MissingConfiguration,
    UnsupportedConfigSectionTypeError,
)
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.api.secure_utils import (
    file_permissions_are_strict,
    windows_get_not_whitelisted_users_with_access,
)
from snowflake.cli.api.utils.types import try_cast_to_bool
from snowflake.connector.compat import IS_WINDOWS
from snowflake.connector.config_manager import CONFIG_MANAGER
from snowflake.connector.constants import CONFIG_FILE, CONNECTIONS_FILE
from snowflake.connector.errors import ConfigSourceError, MissingConfigOptionError
from tomlkit import TOMLDocument, dump
from tomlkit.container import Container
from tomlkit.exceptions import NonExistentKey
from tomlkit.items import Table

log = logging.getLogger(__name__)


class Empty:
    pass


CONNECTIONS_SECTION = "connections"
CLI_SECTION = "cli"
LOGS_SECTION = "logs"
PLUGINS_SECTION = "plugins"

LOGS_SECTION_PATH = [CLI_SECTION, LOGS_SECTION]
PLUGINS_SECTION_PATH = [CLI_SECTION, PLUGINS_SECTION]
PLUGIN_ENABLED_KEY = "enabled"
FEATURE_FLAGS_SECTION_PATH = [CLI_SECTION, "features"]

CONFIG_MANAGER.add_option(
    name=CLI_SECTION,
    parse_str=tomlkit.parse,
    default=dict(),
)


@dataclass
class ConnectionConfig:
    account: Optional[str] = None
    user: Optional[str] = None
    password: Optional[str] = field(default=None, repr=False)
    host: Optional[str] = None
    region: Optional[str] = None
    port: Optional[int] = None
    database: Optional[str] = None
    schema: Optional[str] = None
    warehouse: Optional[str] = None
    role: Optional[str] = None
    authenticator: Optional[str] = None
    private_key_file: Optional[str] = None
    token_file_path: Optional[str] = None

    _other_settings: dict = field(default_factory=lambda: {})

    @classmethod
    def from_dict(cls, config_dict: dict) -> ConnectionConfig:
        known_settings = {}
        other_settings = {}
        for key, value in config_dict.items():
            if key in cls.__dict__:
                known_settings[key] = value
            else:
                other_settings[key] = value
        return cls(**known_settings, _other_settings=other_settings)

    def to_dict_of_known_non_empty_values(self) -> dict:
        return {
            k: v
            for k, v in asdict(self).items()
            if k != "_other_settings" and v is not None
        }

    def _non_empty_other_values(self) -> dict:
        return {k: v for k, v in self._other_settings.items() if v is not None}

    def to_dict_of_all_non_empty_values(self) -> dict:
        return {
            **self.to_dict_of_known_non_empty_values(),
            **self._non_empty_other_values(),
        }


def config_init(config_file: Optional[Path]):
    """
    Initializes the app configuration. Config provided via cli flag takes precedence.
    If config file does not exist we create an empty one.
    """
    from snowflake.cli._app.loggers import create_initial_loggers

    if config_file:
        CONFIG_MANAGER.file_path = config_file
    else:
        _check_default_config_files_permissions()
    if not CONFIG_MANAGER.file_path.exists():
        _initialise_config(CONFIG_MANAGER.file_path)
    _read_config_file()
    create_initial_loggers()


def add_connection_to_proper_file(name: str, connection_config: ConnectionConfig):
    if CONNECTIONS_FILE.exists():
        existing_connections = _read_connections_toml()
        existing_connections.update(
            {name: connection_config.to_dict_of_all_non_empty_values()}
        )
        _update_connections_toml(existing_connections)
        return CONNECTIONS_FILE
    else:
        set_config_value(
            path=[CONNECTIONS_SECTION, name],
            value=connection_config.to_dict_of_all_non_empty_values(),
        )
        return CONFIG_MANAGER.file_path


_DEFAULT_LOGS_CONFIG = {
    "save_logs": True,
    "path": str(CONFIG_MANAGER.file_path.parent / "logs"),
    "level": "info",
}

_DEFAULT_CLI_CONFIG = {LOGS_SECTION: _DEFAULT_LOGS_CONFIG}


@contextmanager
def _config_file():
    _read_config_file()
    conf_file_cache = CONFIG_MANAGER.conf_file_cache
    yield conf_file_cache
    _dump_config(conf_file_cache)


def _read_config_file():
    with warnings.catch_warnings():
        if IS_WINDOWS:
            warnings.filterwarnings(
                action="ignore",
                message="Bad owner or permissions.*",
                module="snowflake.connector.config_manager",
            )

            if not file_permissions_are_strict(CONFIG_MANAGER.file_path):
                users = ", ".join(
                    windows_get_not_whitelisted_users_with_access(
                        CONFIG_MANAGER.file_path
                    )
                )
                warnings.warn(
                    f"Unauthorized users ({users}) have access to configuration file {CONFIG_MANAGER.file_path}.\n"
                    f'Run `icacls "{CONFIG_MANAGER.file_path}" /remove:g <USER_ID>` on those users to restrict permissions.'
                )

        try:
            CONFIG_MANAGER.read_config()
        except ConfigSourceError as exception:
            raise ClickException(
                f"Configuration file seems to be corrupted. {str(exception.__cause__)}"
            )


def _initialise_logs_section():
    with _config_file() as conf_file_cache:
        if conf_file_cache.get(CLI_SECTION) is None:
            conf_file_cache[CLI_SECTION] = _DEFAULT_CLI_CONFIG
        if conf_file_cache[CLI_SECTION].get(LOGS_SECTION) is None:
            conf_file_cache[CLI_SECTION][LOGS_SECTION] = _DEFAULT_LOGS_CONFIG


def set_config_value(path: List[str], value: Any) -> None:
    """Sets value in config.
    For example to set value "val" to key "key" in section [a.b.c], call
    set_config_value(["a", "b", "c", "key"], "val").
    If you want to override a whole section, value should be a dictionary.
    """
    with _config_file() as conf_file_cache:
        current_config_dict = conf_file_cache
        for key in path[:-1]:
            if key not in current_config_dict:
                current_config_dict[key] = {}
            current_config_dict = current_config_dict[key]
        current_config_dict[path[-1]] = value


def get_logs_config() -> dict:
    logs_config = _DEFAULT_LOGS_CONFIG.copy()
    if config_section_exists(*LOGS_SECTION_PATH):
        logs_config.update(**get_config_section(*LOGS_SECTION_PATH))
    return logs_config


def get_plugins_config() -> dict:
    if config_section_exists(*PLUGINS_SECTION_PATH):
        return get_config_section(*PLUGINS_SECTION_PATH)
    else:
        return {}


def connection_exists(connection_name: str) -> bool:
    return config_section_exists(CONNECTIONS_SECTION, connection_name)


def config_section_exists(*path) -> bool:
    try:
        _find_section(*path)
        return True
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        return False


def get_all_connections() -> dict[str, ConnectionConfig]:
    return {
        k: ConnectionConfig.from_dict(connection_dict)
        for k, connection_dict in get_config_section("connections").items()
    }


def get_connection_dict(connection_name: str) -> dict:
    try:
        return get_config_section(CONNECTIONS_SECTION, connection_name)
    except KeyError:
        raise MissingConfiguration(f"Connection {connection_name} is not configured")


def get_default_connection_name() -> str:
    return CONFIG_MANAGER["default_connection_name"]


def get_default_connection_dict() -> dict:
    def_connection_name = get_default_connection_name()
    if not connection_exists(def_connection_name):
        raise MissingConfiguration(
            f"Couldn't find connection for default connection `{def_connection_name}`. "
            f"Specify connection name or configure default connection."
        )
    return get_connection_dict(def_connection_name)


def get_config_section(*path) -> dict:
    section = _find_section(*path)
    if isinstance(section, Container):
        return {s: _merge_section_with_env(section[s], *path, s) for s in section}
    if isinstance(section, dict):
        return _merge_section_with_env(section, *path)
    raise UnsupportedConfigSectionTypeError(type(section))


def get_config_value(*path, key: str, default: Optional[Any] = Empty) -> Any:
    """Looks for given key under nested path in toml file."""
    env_variable = get_env_value(*path, key=key)
    if env_variable:
        return env_variable
    try:
        return get_config_section(*path)[key]
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        if default is not Empty:
            return default
        raise


def get_config_bool_value(*path, key: str, default: Optional[bool]) -> Optional[bool]:
    value = get_config_value(*path, key=key, default=None)

    if value is None:
        return default

    try:
        return try_cast_to_bool(value)
    except ValueError:
        raise ClickException(
            f"Expected boolean value for {'.'.join((*path, key))} option."
        )


def _initialise_config(config_file: Path) -> None:
    config_file = SecurePath(config_file)
    config_file.parent.mkdir(parents=True, exist_ok=True)
    config_file.touch()
    _initialise_logs_section()
    log.info("Created Snowflake configuration file at %s", CONFIG_MANAGER.file_path)


def get_env_variable_name(*path, key: str) -> str:
    return ("_".join(["snowflake", *path, key])).upper()


def get_env_value(*path, key: str) -> str | None:
    return os.environ.get(get_env_variable_name(*path, key=key))


def _find_section(*path) -> TOMLDocument:
    section = CONFIG_MANAGER
    idx = 0
    while idx < len(path):
        section = section[path[idx]]
        idx += 1
    return section


def _merge_section_with_env(section: Union[Table, Any], *path) -> Dict[str, str]:
    if isinstance(section, Table):
        env_variables = _get_envs_for_path(*path)
        section_copy = section.copy()
        section_copy.update(env_variables)
        return section_copy.unwrap()
    # It's a atomic value
    return section


def _get_envs_for_path(*path) -> dict:
    env_variables_prefix = "_".join(["SNOWFLAKE"] + [p.upper() for p in path]) + "_"
    return {
        k.replace(env_variables_prefix, "").lower(): os.environ[k]
        for k in os.environ.keys()
        if k.startswith(env_variables_prefix)
    }


def _dump_config(config_and_connections: Dict):
    config_toml_dict = config_and_connections.copy()

    if CONNECTIONS_FILE.exists():
        # update connections in connections.toml
        # it will add only connections (maybe updated) which were originally read from connections.toml
        # it won't add connections from config.toml
        # because config manager doesn't have connections from config.toml if connections.toml exists
        _update_connections_toml(config_and_connections.get("connections") or {})
        # to config.toml save only connections from config.toml
        connections_to_save_in_config_toml = _read_config_file_toml().get("connections")
        if connections_to_save_in_config_toml:
            config_toml_dict["connections"] = connections_to_save_in_config_toml
        else:
            config_toml_dict.pop("connections", None)

    with SecurePath(CONFIG_MANAGER.file_path).open("w+") as fh:
        dump(config_toml_dict, fh)


def _check_default_config_files_permissions() -> None:
    if not IS_WINDOWS:
        if CONNECTIONS_FILE.exists() and not file_permissions_are_strict(
            CONNECTIONS_FILE
        ):
            raise ConfigFileTooWidePermissionsError(CONNECTIONS_FILE)
        if CONFIG_FILE.exists() and not file_permissions_are_strict(CONFIG_FILE):
            raise ConfigFileTooWidePermissionsError(CONFIG_FILE)


from typing import Literal


def get_feature_flags_section() -> Dict[str, bool | Literal["UNKNOWN"]]:
    if not config_section_exists(*FEATURE_FLAGS_SECTION_PATH):
        return {}

    flags = get_config_section(*FEATURE_FLAGS_SECTION_PATH)

    def _bool_or_unknown(value):
        try:
            return try_cast_to_bool(value)
        except ValueError:
            return "UNKNOWN"

    return {k: _bool_or_unknown(v) for k, v in flags.items()}


def _read_config_file_toml() -> dict:
    return tomlkit.loads(CONFIG_MANAGER.file_path.read_text()).unwrap()


def _read_connections_toml() -> dict:
    return tomlkit.loads(CONNECTIONS_FILE.read_text()).unwrap()


def _update_connections_toml(connections: dict):
    with open(CONNECTIONS_FILE, "w") as f:
        f.write(tomlkit.dumps(connections))
