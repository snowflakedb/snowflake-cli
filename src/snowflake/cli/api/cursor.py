from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.exceptions import CliSqlError
from snowflake.connector.cursor import DictCursor, SnowflakeCursor
from snowflake.connector.errors import ProgrammingError


# todo: remove usage of custom cursor & handle resolution in SnowTyper.exception_handler
class ExtendedExitCodes:
    def execute(self, command, *args, **kwargs):
        try:
            super().execute(command, *args, **kwargs)
        except ProgrammingError as pex:
            raise pex
            if get_cli_context().enhanced_exit_codes:
                raise CliSqlError(pex.msg) from pex
            raise pex


class CliSnowflakeCursor(ExtendedExitCodes, SnowflakeCursor):
    """Cli Cursor with override of execution for custom exception handling."""


class CliDictCursor(ExtendedExitCodes, DictCursor):
    """Cli Dict Cursor with override of execution for custom exception handling."""
