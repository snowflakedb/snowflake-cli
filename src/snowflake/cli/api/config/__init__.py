# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Configuration package with support for legacy and next-generation config systems.

This package provides a feature flag mechanism to switch between the legacy
configuration system and the new config-ng system using the environment variable
SNOWFLAKE_CLI_CONFIG_NG.

Usage:
    # Enable new config system
    export SNOWFLAKE_CLI_CONFIG_NG=1

    # Use legacy config system (default)
    unset SNOWFLAKE_CLI_CONFIG_NG

Environment Variables:
    SNOWFLAKE_CLI_CONFIG_NG: Set to '1', 'true', 'yes', or 'on' to enable config-ng
"""

from __future__ import annotations

import os
from typing import Literal

# Environment variable to enable the new config-ng system
CONFIG_NG_ENV_VAR: Literal["SNOWFLAKE_CLI_CONFIG_NG"] = "SNOWFLAKE_CLI_CONFIG_NG"


def _is_config_ng_enabled() -> bool:
    """
    Check if the new config-ng system should be used.

    Returns:
        True if config-ng should be used, False for legacy system
    """
    return os.environ.get(CONFIG_NG_ENV_VAR, "").lower() in ("1", "true", "yes", "on")


# Conditionally import from legacy or config-ng based on environment variable
if _is_config_ng_enabled():
    # Import ONLY from the new config-ng system
    # This will fail if config-ng is not fully implemented, which is intentional
    # to help identify what needs to be implemented
    try:
        from .config_ng import *  # noqa: F403, F401
    except ImportError as e:
        # If config-ng is not fully implemented, fail fast with a clear error
        raise ImportError(
            f"Config-ng is enabled via {CONFIG_NG_ENV_VAR} but is not fully implemented. "
            f"Missing implementation: {e}. "
            f"Either implement the missing functionality in config_ng.py or "
            f"disable config-ng by unsetting {CONFIG_NG_ENV_VAR}."
        ) from e
else:
    # Use the legacy configuration system
    from .legacy import *  # noqa: F403, F401

# Export the feature flag utilities and factory method for external use
__all__ = ["_is_config_ng_enabled", "CONFIG_NG_ENV_VAR", "get_config_manager"]
