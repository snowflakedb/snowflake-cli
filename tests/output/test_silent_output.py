# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import pytest
from snowflake.cli.api.cli_global_context import get_cli_context


@pytest.mark.parametrize(
    "command, expected_value",
    (
        pytest.param(("sql"), False, id="silent is False"),
        pytest.param(("sql", "--silent"), True, id="silent is True"),
    ),
)
def test_silent_in_global_context(
    command: tuple[str, ...],
    expected_value: bool,
    runner,
):
    runner.invoke(command)

    assert get_cli_context().silent is expected_value


def test_silent_output_help(runner, snapshot):
    result = runner.invoke(["streamlit", "get-url", "--help"], catch_exceptions=False)

    assert result.exit_code == 0, result.output
    snapshot.assert_match(result.output)


def test_proper_context_values_for_silent(runner):
    result = runner.invoke(["streamlit", "get-url", "--silent", "--help"])
    assert runner.app is not None
    assert runner.app

    assert result.exit_code == 0, result.output
