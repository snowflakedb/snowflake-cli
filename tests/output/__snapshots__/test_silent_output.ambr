# serializer version: 1
# name: test_silent_output_help
  '''
                                                                                  
   Usage: root streamlit get-url [OPTIONS] NAME                                   
                                                                                  
   Returns a URL to the specified Streamlit app                                   
                                                                                  
  +- Arguments ------------------------------------------------------------------+
  | *    name      TEXT  Identifier of the Streamlit app; for example:           |
  |                      my_streamlit                                            |
  |                      [required]                                              |
  +------------------------------------------------------------------------------+
  +- Options --------------------------------------------------------------------+
  | --open            Whether to open the Streamlit app in a browser.            |
  | --help  -h        Show this message and exit.                                |
  +------------------------------------------------------------------------------+
  +- Connection configuration ---------------------------------------------------+
  | --connection,--environment     -c      TEXT     Name of the connection, as   |
  |                                                 defined in your config.toml  |
  |                                                 file. Default: default.      |
  | --host                                 TEXT     Host address for the         |
  |                                                 connection. Overrides the    |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --port                                 INTEGER  Port for the connection.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --account,--accountname                TEXT     Name assigned to your        |
  |                                                 Snowflake account. Overrides |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --user,--username                      TEXT     Username to connect to       |
  |                                                 Snowflake. Overrides the     |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --password                             TEXT     Snowflake password.          |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --authenticator                        TEXT     Snowflake authenticator.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --private-key-file,--private…          TEXT     Snowflake private key file   |
  |                                                 path. Overrides the value    |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --token-file-path                      TEXT     Path to file with an OAuth   |
  |                                                 token to use when connecting |
  |                                                 to Snowflake.                |
  | --database,--dbname                    TEXT     Database to use. Overrides   |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --schema,--schemaname                  TEXT     Database schema to use.      |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --role,--rolename                      TEXT     Role to use. Overrides the   |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --warehouse                            TEXT     Warehouse to use. Overrides  |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --temporary-connection         -x               Uses a connection defined    |
  |                                                 with command line            |
  |                                                 parameters, instead of one   |
  |                                                 defined in config            |
  | --mfa-passcode                         TEXT     Token to use for             |
  |                                                 multi-factor authentication  |
  |                                                 (MFA)                        |
  | --enable-diag                                   Whether to generate a        |
  |                                                 connection diagnostic        |
  |                                                 report.                      |
  | --diag-log-path                        TEXT     Path for the generated       |
  |                                                 report. Defaults to system   |
  |                                                 temporary directory.         |
  | --diag-allowlist-path                  TEXT     Path to a JSON file that     |
  |                                                 contains allowlist           |
  |                                                 parameters.                  |
  | --oauth-client-id                      TEXT     OAuth client ID to use when  |
  |                                                 connecting to Snowflake.     |
  | --oauth-client-secret                  TEXT     OAuth client secret to use   |
  |                                                 when connecting to           |
  |                                                 Snowflake.                   |
  | --oauth-authorization-url              TEXT     OAuth authorization URL to   |
  |                                                 use when connecting to       |
  |                                                 Snowflake.                   |
  | --oauth-redirect-uri                   TEXT     OAuth redirect URI to use    |
  |                                                 when connecting to           |
  |                                                 Snowflake.                   |
  | --oauth-scope                          TEXT     OAuth scope to use when      |
  |                                                 connecting to Snowflake.     |
  | --oauth-security-features              TEXT     OAuth security features to   |
  |                                                 use when connecting to       |
  |                                                 Snowflake. Expects a         |
  |                                                 comma-separated list of      |
  |                                                 values.                      |
  | --client-store-temporary-cre…                   Store the temporary          |
  |                                                 credential.                  |
  +------------------------------------------------------------------------------+
  +- Global configuration -------------------------------------------------------+
  | --format                       [TABLE|JSON]  Specifies the output format.    |
  |                                              [default: TABLE]                |
  | --verbose              -v                    Displays log entries for log    |
  |                                              levels info and higher.         |
  | --debug                                      Displays log entries for log    |
  |                                              levels debug and higher; debug  |
  |                                              logs contain additional         |
  |                                              information.                    |
  | --silent                                     Turns off intermediate output   |
  |                                              to console.                     |
  | --enhanced-exit-codes                        Differentiate exit error codes  |
  |                                              based on failure type.          |
  |                                              [env var:                       |
  |                                              SNOWFLAKE_ENHANCED_EXIT_CODES]  |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
